### PR TITLE
Feature/aa 374 add accommodation change history part2

### PR DESCRIPTION
**Pull request by hasmikharutyunyan**

### CODE REVIEW

File: HappyTravel.Nakijin.AccommodationChangeLogger/Models/DuplicateAccommodationSupplierCodesMergeEventData.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+{
+    public readonly struct DuplicateAccommodationSupplierCodesMergeEventData
+    {
+        public DuplicateAccommodationSupplierCodesMergeEventData(int duplicateAccommodationId, Dictionary<Suppliers, string> sourceAccommodationSupplierCodes,
+            Dictionary<Suppliers, string> duplicateAccommodationSupplierCodes)
+        {
+            DuplicateAccommodationId = duplicateAccommodationId;
+            SourceAccommodationSupplierCodes = sourceAccommodationSupplierCodes;
+            DuplicateAccommodationSupplierCodes = duplicateAccommodationSupplierCodes;
+
```
Comment (oleg1std) [inline HappyTravel.Nakijin.AccommodationChangeLogger/Models/DuplicateAccommodationSupplierCodesMergeEventData.cs:14]: extra line

File: HappyTravel.Nakijin.AccommodationChangeLogger/Models/NewSupplierAdditionEventData.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+{
+    public readonly struct NewSupplierAdditionEventData
+    {
+        public NewSupplierAdditionEventData(Dictionary<Suppliers, string> existingAccommodationSupplierCodes, Suppliers supplier, string supplierAccommodationCode)
+        {
+            ExistingAccommodationSupplierCodes = existingAccommodationSupplierCodes;
+            Supplier = supplier;
+            SupplierAccommodationCode = supplierAccommodationCode;
+        }
+
```
Comment (oleg1std) [inline HappyTravel.Nakijin.AccommodationChangeLogger/Models/NewSupplierAdditionEventData.cs:14]: extra line needed

File: HappyTravel.Nakijin.AccommodationChangeLogger/Models/DuplicateAccommodationSupplierCodesMergeEventData.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+        }
+
```
Comment (oleg1std) [inline HappyTravel.Nakijin.AccommodationChangeLogger/Models/DuplicateAccommodationSupplierCodesMergeEventData.cs:16]: extra line is needed

File: HappyTravel.Nakijin.AccommodationChangeLogger/Models/DuplicateAccommodationCommonSuppliersCodesRemoveEventData.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+{
+    public readonly struct DuplicateAccommodationCommonSuppliersCodesRemoveEventData
+    {
+        public DuplicateAccommodationCommonSuppliersCodesRemoveEventData(int sourceAccommodationId, Dictionary<Suppliers, string> duplicateAccommodationSupplierCodes, Dictionary<Suppliers, string> sourceAccommodationSupplierCodes)
+        {
+            SourceAccommodationId = sourceAccommodationId;
+            DuplicateAccommodationSupplierCodes = duplicateAccommodationSupplierCodes;
+            SourceAccommodationSupplierCodes = sourceAccommodationSupplierCodes;
+        }
+
```
Comment (oleg1std) [inline HappyTravel.Nakijin.AccommodationChangeLogger/Models/DuplicateAccommodationCommonSuppliersCodesRemoveEventData.cs:14]: extra line is needed

File: HappyTravel.Nakijin.AccommodationChangeLogger/Models/SupplierCodeRemoveEventData.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    public readonly struct SupplierCodeRemoveEventData
+    {
+        public SupplierCodeRemoveEventData(Suppliers supplier, string supplierCode, Dictionary<Suppliers, string> accommodationSupplierCodes, SupplierCodeRemoveReasons reason)
+        {
+            Supplier = supplier;
+            SupplierCode = supplierCode;
+            AccommodationSupplierCodes = accommodationSupplierCodes;
+            Reason = reason;
+        }
+
```
Comment (oleg1std) [inline HappyTravel.Nakijin.AccommodationChangeLogger/Models/SupplierCodeRemoveEventData.cs:16]: please checl the formatting

File: HappyTravel.Nakijin.AccommodationChangeLogger/Services/AccommodationChangeLogService.cs [hunk 6]
```diff
@@ -21,22 +24,76 @@ public AccommodationChangeLogService(NakijinContext context)
...
+            return AddAccommodationChangeLog(duplicateAccommodationSupplierCodesMergeLog);
+        }
+
+
+        public Task AddDuplicateAccommodationCommonSuppliersCodesRemoveEvent(int accommodationId, AccommodationChangeEventSources source,
+            int sourceAccommodationId, Dictionary<Suppliers, string> duplicateAccommodationSupplierCodes,
+            Dictionary<Suppliers, string> sourceAccommodationSupplierCodes)
+        {
+            var duplicateAccommodationCommonSuppliersCodesRemoveLog = AccommodationChangeLogConverter.GetDuplicateAccommodationCommonSuppliersCodesRemoveLog(
+                accommodationId,
```
Comment (oleg1std) [inline HappyTravel.Nakijin.AccommodationChangeLogger/Services/AccommodationChangeLogService.cs:52]: All the argument could be fit in one line

File: HappyTravel.Nakijin.AccommodationChangeLogger/Services/IAccommodationChangeLogCollectorService.cs [hunk 7]
```diff
@@ -13,23 +14,33 @@ public interface IAccommodationChangeLogCollectorService
...
+        void CollectNewSupplierAdditionEvent(int accommodationId, AccommodationChangeEventSources source,
+            Dictionary<Suppliers, string> existingAccommodationSupplierCodes, Suppliers supplier, string supplierAccommodationCode);
+        
+        void CollectDuplicateAccommodationSupplierCodesMergeEvent(int accommodationId, AccommodationChangeEventSources source, int duplicateAccommodationId,
+            Dictionary<Suppliers, string> sourceAccommodationSupplierCodes, Dictionary<Suppliers, string> duplicateAccommodationSupplierCodes);
+      
+        void CollectDuplicateAccommodationCommonSuppliersCodesRemoveEvent(int accommodationId, AccommodationChangeEventSources source,
+            int sourceAccommodationId, Dictionary<Suppliers, string> duplicateAccommodationSupplierCodes,
+            Dictionary<Suppliers, string> sourceAccommodationSupplierCodes);
+        void CollectRemoveEvent(int accommodationId, AccommodationChangeEventSources source, AccommodationRemoveReasons reason);
```
Comment (oleg1std) [inline HappyTravel.Nakijin.AccommodationChangeLogger/Services/IAccommodationChangeLogCollectorService.cs:40]: A line is missed

File: HappyTravel.Nakijin.AccommodationChangeLogger/Converters/AccommodationChangeLogConverter.cs [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+using HappyTravel.SuppliersCatalog;
+using Newtonsoft.Json;
+
+namespace HappyTravel.Nakijin.AccommodationChangeLogger.Converters
+{
+    public static class AccommodationChangeLogConverter
+    {
+        public static AccommodationChangeLog GetActivationChangeLog(int accommodationId, AccommodationChangeEventSources eventSource)
+        {
+            var activationEventData = new ActivationEventData();
```
Comment (AlexeyShok) [inline HappyTravel.Nakijin.AccommodationChangeLogger/Converters/AccommodationChangeLogConverter.cs:18]: We usually insert a blank line before the return line from the method

File: HappyTravel.Nakijin.AccommodationChangeLogger/Converters/AccommodationChangeLogConverter.cs [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+        private static AccommodationChangeLog GetAccommodationChangeLog<T>(int accommodationId, AccommodationChangeEventSources eventSource,
+            AccommodationChangeEventTypes eventType, T eventData)
+            => new AccommodationChangeLog
+            {
+                AccommodationId = accommodationId,
+                EventSource = eventSource,
+                EventType = eventType,
+                ChangeEventData = JsonDocument.Parse(JsonConvert.SerializeObject(eventData)),
+                Created = DateTime.UtcNow
```
Comment (AlexeyShok) [inline HappyTravel.Nakijin.AccommodationChangeLogger/Converters/AccommodationChangeLogConverter.cs:111]: For further unit test coverage, it is better to use the DateTimeProvider method call in such cases instead DateTime.UtcNow

File: HappyTravel.Nakijin.AccommodationChangeLogger/Models/ActivationEventData.cs [hunk 10]
```diff
@@ -1,16 +1,7 @@
...
     public readonly struct ActivationEventData
     {
-        public ActivationEventData(Dictionary<Suppliers, string> supplierAccommodationCodes)
-        {
-            SupplierAccommodationCodes = supplierAccommodationCodes;
-        }
-
-
-        public Dictionary<Suppliers, string> SupplierAccommodationCodes { get; }
+
```
Comment (AlexeyShok) [inline HappyTravel.Nakijin.AccommodationChangeLogger/Models/ActivationEventData.cs:5]: Extra empty line

File: HappyTravel.Nakijin.AccommodationChangeLogger/Services/AccommodationChangeLogService.cs [hunk 11]
```diff
@@ -50,20 +107,38 @@ public Task AddChangeEventsList(List<AccommodationChangeLog> changeLogs, Cancell
...
+                AccommodationChangeLogConverter.GetDeactivationChangeLog(accommodationId, source, deactivationReason, deactivationReasonDescription);
             return AddAccommodationChangeLog(deactivationLog);
         }
 
 
+        public Task AddDeactivationEvents(List<int> accommodationIds, AccommodationChangeEventSources source, DeactivationReasons deactivationReason,
+            string? deactivationReasonDescription)
+        {
+            var deactivationLogs = accommodationIds.Select(id
+                => AccommodationChangeLogConverter.GetDeactivationChangeLog(id, source, deactivationReason, deactivationReasonDescription)).ToList();
```
Comment (AlexeyShok) [inline HappyTravel.Nakijin.AccommodationChangeLogger/Services/AccommodationChangeLogService.cs:123]: Move ToList () to a new line is more clearly

File: HappyTravel.Nakijin.AccommodationChangeLogger/Converters/AccommodationChangeLogConverter.cs (hunk 9 above)
Comment (hasmikharutyunyan) [reply HappyTravel.Nakijin.AccommodationChangeLogger/Converters/AccommodationChangeLogConverter.cs:111]: discussed

**Review by oleg1std — APPROVED** [review]

**Pull request merged**